srml/contract/src/lib.rs

```diff
@@ -14,41 +14,82 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate. If not, see <http://www.gnu.org/licenses/>.
 
-//! Smart-contract module for runtime; Allows deployment and execution of smart-contracts
-//! expressed in WebAssembly.
+//! # Contract Module
+//! 
+//! The contract module provides functionality for the runtime to deploy and execute WebAssembly smart-contracts.
+//! The supported dispatchable functions are documented as part of the [`Call`](./enum.Call.html) enum.
+//! 
+//! ## Overview
+//! 
+//! This module extends accounts (see `Balances` module) to have smart-contract functionality.
+//! These "smart-contract accounts" have the ability to create smart-contracts and make calls to other contract 
+//! and non-contract accounts.
+//! 
+//! The smart-contract code is stored once in a code cache, and later retrievable via a code hash.
+//! This means that multiple smart-contracts can be instantiated from the same code cache, without replicating 
+//! the code each time.
+//! 
+//! When a smart-contract is called, its associated code is retrieved via the code hash and gets executed.
+//! This call can alter the storage entries of the smart-contract account, create new smart-contracts,
+//! or call other smart-contracts.
+//! 
+//! Finally, when the `balances` module determines an account is dead (i.e. account balance fell below the
+//! existential deposit), it reaps the account. This will delete the associated code and storage of the 
+//! smart-contract account.
+//! 
+//! **NOTE:** Call failures are not always cascading. For example, if contract A calls contract B, and B
```

Comment:
This reads really weird now. Like out of nowhere you are talking about call failures. Is this actually information we need? If so, should it be part of a new custom section like `Scenarios` or something?

Reply:
Sure , splitting it out into its own `notable scenarios` section, as the content here will grow in the future. We'll also need a `security considerations` section in the future as coding patterns become clearer.